handler/plchandler/plchandler.go

```diff
@@ -54,84 +71,125 @@ func NewPLCHandler(core handler.CoreHandler) *PLCHandler {
 	h.SlaveId = 1
 	h.Timeout = 2 * time.Second
 
-	rv := &PLCHandler{ClientHandler: h}
+	rv := &PLCHandler{clientHandler: h}
 	rv.Init()
 	rv.Name = "plchandler"
 	rv.CoreHandler = core
 	return rv
 }
 
+func makePollingMemBlocks(tagsMemBlocks []MemBlock) []MemBlock {
+	if len(tagsMemBlocks) == 0 {
+		return tagsMemBlocks
+	}
+
+	rv := []MemBlock{}
+
+	sort.Slice(tagsMemBlocks, func(i, j int) bool {
+		return tagsMemBlocks[i].address < tagsMemBlocks[j].address
+	})
+
+	block := MemBlock{tags: []*Tag{}}
+	for _, mb := range tagsMemBlocks {
+		if block.size == 0 {
+			block.address = mb.address
+			block.size = mb.size
+			block.tags = append(block.tags, mb.tags[0])
+			continue
+		}
+		newBlockSize := mb.address + mb.size - block.address
+		breakSize := mb.address - block.address - block.size
+		if newBlockSize > maxBlockSize || breakSize > maxBreakSize {
+			rv = append(rv, block)
+			block = MemBlock{address: mb.address, size: mb.size}
+			block.tags = append(block.tags, mb.tags[0])
+			continue
+		}
+		block.size = newBlockSize
+		block.tags = append(block.tags, mb.tags[0])
+	}
+	if block.size > 0 {
+		rv = append(rv, block)
+	}
+
+	return rv
+}
+
 func (ph *PLCHandler) Start() error {
 	ph.BaseHandler.Start()
 
-	ph.Values = make(map[string]string)
+	ph.tags = make(map[string]*Tag)
 
 	tags := getTagsConfig()
+	tagsCount := len(tags.Input) + len(tags.Output)
+	tagsMemBlocks := make([]MemBlock, 0, tagsCount)
 	for _, tag := range tags.Input {
-		ph.Tags = append(ph.Tags, tag)
+		tag.Size = 1
+		tag.Type = typeInput
+		ph.tags[tag.Name] = &tag
+		mb := MemBlock{address: tag.Address, size: tag.Size, tags: []*Tag{&tag}}
+		tagsMemBlocks = append(tagsMemBlocks, mb)
 	}
 	for _, tag := range tags.Output {
-		ph.Tags = append(ph.Tags, tag)
+		tag.Size = 1
+		tag.Type = typeOutput
+		ph.tags[tag.Name] = &tag
+		mb := MemBlock{address: tag.Address, size: tag.Size, tags: []*Tag{&tag}}
+		tagsMemBlocks = append(tagsMemBlocks, mb)
 	}
-	fmt.Printf("%+v\n", ph.Tags)
+	ph.pollingMemBlocks = makePollingMemBlocks(tagsMemBlocks)
 
-	err := ph.ClientHandler.Connect()
+	err := ph.clientHandler.Connect()
 	if err != nil {
 		return err
 	}
-	defer ph.ClientHandler.Close()
-	//
-	client := modbus.NewClient(ph.ClientHandler)
+	defer ph.clientHandler.Close()
+
+	client := modbus.NewClient(ph.clientHandler)
+	return ph.loop(&client)
+}
+
+func (ph *PLCHandler) loop(client *modbus.Client) error {
```

Comment:
modbus.Client is interface, you should not use pointer to interface

Reply:
fixed